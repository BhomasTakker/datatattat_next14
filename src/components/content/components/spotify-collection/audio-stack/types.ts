import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
