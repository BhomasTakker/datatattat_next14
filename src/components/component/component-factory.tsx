import {
	ComponentsMap,
	ComponentsOptions,
} from "@/components/component/component-map";
import { PageComponent } from "@/types/page";

export const ComponentFactory = ({
	component,
}: {
	component: PageComponent;
}) => {
	const { componentType } = component || {};

	const Component = ComponentsMap.get(componentType as ComponentsOptions);

	// do better error handling
	if (!Component) {
		return <div>Component not found</div>;
	}

	return <Component component={component} />;
};
