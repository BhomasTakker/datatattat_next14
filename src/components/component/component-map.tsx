import { PageComponent } from "@/types/page";

const ArticleCollection = ({ component }: { component: PageComponent }) => {
	console.log({ component });
	return <div>Article Collection</div>;
};

type Components = typeof ArticleCollection;

export enum ComponentsOptions {
	ArticleCollection = "ArticleCollection",
}

export const ComponentsMap = new Map<ComponentsOptions, Components>([
	[ComponentsOptions.ArticleCollection, ArticleCollection],
]);
