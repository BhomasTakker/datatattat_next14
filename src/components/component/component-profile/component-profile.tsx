export const ComponentProfile = ({ profile }: { profile: object }) => {
	console.log({ profile });
	return <h2>Component Title</h2>;
};
