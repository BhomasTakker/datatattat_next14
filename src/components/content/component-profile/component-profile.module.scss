@use "../../../scss/abstracts/font" as *;
@use "../../../scss/abstracts/colours" as colours;
@use "../../../scss/abstracts/units" as *;

.root {
	container: profile-container / inline-size;
	display: flex;

	.title {
		font: $titleXs;
		color: colours.$titleColourDark;

		// we cannot use variables with container queries
		// I think this is the case / follow up on this
		@container profile-container (min-width: 640px) {
			font: $titleSm;
		}

		@container profile-container (min-width: 1024px) {
			font: $titleMd;
		}
	}
}
