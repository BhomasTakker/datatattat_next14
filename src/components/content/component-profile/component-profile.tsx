import Link from "next/link";
import { ComponentProfileProps } from "../../../types/page";
import styles from "./component-profile.module.scss";

export const ComponentProfile = ({
	profile,
}: {
	profile: ComponentProfileProps;
}) => {
	const { componentTitle, showComponentTitle, componentTitleLink } =
		profile || {};

	const titleComponent = <h2 className={styles.title}>{componentTitle}</h2>;
	const titleComponentToRender = componentTitleLink ? (
		<Link href={componentTitleLink}>{titleComponent}</Link>
	) : (
		titleComponent
	);

	if (!showComponentTitle) {
		return <></>;
	}

	return (
		<div className={styles.root}>
			{showComponentTitle && componentTitle && titleComponentToRender}
		</div>
	);
};
