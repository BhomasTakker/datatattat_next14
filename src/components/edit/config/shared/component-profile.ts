import { GenericInput, InputListProps } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../inputs/inputs";

const componentProfileOptions = (): GenericInput[] => [
	{
		id: "componentTitle",
		type: EditInputs.text,
		label: "Component Title",
		required: false,
	},
	{
		id: "componentTitleLink",
		type: EditInputs.text,
		label: "Component Title Link",
		required: false,
	},
];

/**
 * Add to componentProps object
 */
export const componentProfile: InputListProps = {
	id: "componentProps",
	type: EditInputs.inputList,
	label: "Component",
	createObject: true,
	inputs: componentProfileOptions(),
};
