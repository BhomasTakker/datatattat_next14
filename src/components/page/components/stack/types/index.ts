import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
