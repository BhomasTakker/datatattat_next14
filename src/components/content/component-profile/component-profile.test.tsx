import React from "react";
import { render, screen } from "@testing-library/react";
import { ComponentProfile } from "./component-profile";
import { ComponentProfileProps } from "@/types/component";
import "@testing-library/jest-dom";

// Mock next/link for testing
jest.mock("next/link", () => {
	return ({ children, href }: any) => <a href={href}>{children}</a>;
});

const baseProfile: ComponentProfileProps = {
	componentTitle: "Test Title",
	showComponentTitle: true,
	// @ts-expect-error mock data
	componentTitleLink: undefined,
};

describe("ComponentProfile", () => {
	it("renders the component title when provided", () => {
		render(<ComponentProfile profile={baseProfile} />);
		expect(screen.getByText("Test Title")).toBeInTheDocument();
	});

	it("renders the component title as a link when componentTitleLink is provided", () => {
		const profile: ComponentProfileProps = {
			...baseProfile,
			componentTitleLink: "/test-link",
		};
		render(<ComponentProfile profile={profile} />);
		const link = screen.getByRole("link", { name: "Test Title" });
		expect(link).toBeInTheDocument();
		expect(link).toHaveAttribute("href", "/test-link");
	});

	it("does not render anything if componentTitle is undefined", () => {
		const profile: ComponentProfileProps = {
			...baseProfile,
			// @ts-expect-error mock data
			componentTitle: undefined,
		};
		const { container } = render(<ComponentProfile profile={profile} />);
		expect(container).toBeEmptyDOMElement();
	});

	it("does not render anything if componentTitle is an empty string", () => {
		const profile: ComponentProfileProps = {
			...baseProfile,
			componentTitle: "",
		};
		const { container } = render(<ComponentProfile profile={profile} />);
		expect(container).toBeEmptyDOMElement();
	});

	it("renders with custom class names from styles", () => {
		render(<ComponentProfile profile={baseProfile} />);
		const title = screen.getByText("Test Title");
		expect(title).toHaveClass("title");
	});

	it("renders the root div with the correct class", () => {
		render(<ComponentProfile profile={baseProfile} />);
		const rootDiv = screen.getByText("Test Title").closest("div");
		expect(rootDiv).toHaveClass("root");
	});

	describe("Snaps", () => {
		it("matches the snapshot with a title", () => {
			const { asFragment } = render(<ComponentProfile profile={baseProfile} />);
			expect(asFragment()).toMatchSnapshot();
		});

		it("matches the snapshot with a title link", () => {
			const profile: ComponentProfileProps = {
				...baseProfile,
				componentTitleLink: "/test-link",
			};
			const { asFragment } = render(<ComponentProfile profile={profile} />);
			expect(asFragment()).toMatchSnapshot();
		});

		it("matches the snapshot when componentTitle is undefined", () => {
			const profile: ComponentProfileProps = {
				...baseProfile,
				// @ts-expect-error mock data
				componentTitle: undefined,
			};
			const { asFragment } = render(<ComponentProfile profile={profile} />);
			expect(asFragment()).toMatchSnapshot();
		});

		it("matches the snapshot when componentTitle is an empty string", () => {
			const profile: ComponentProfileProps = {
				...baseProfile,
				componentTitle: "",
			};
			const { asFragment } = render(<ComponentProfile profile={profile} />);
			expect(asFragment()).toMatchSnapshot();
		});
	});
});
