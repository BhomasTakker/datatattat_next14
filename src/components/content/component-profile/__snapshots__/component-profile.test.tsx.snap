// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComponentProfile Snaps matches the snapshot when componentTitle is an empty string 1`] = `<DocumentFragment />`;

exports[`ComponentProfile Snaps matches the snapshot when componentTitle is undefined 1`] = `<DocumentFragment />`;

exports[`ComponentProfile Snaps matches the snapshot with a title 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <h2
      class="title"
    >
      Test Title
    </h2>
  </div>
</DocumentFragment>
`;

exports[`ComponentProfile Snaps matches the snapshot with a title link 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <a
      href="/test-link"
    >
      <h2
        class="title"
      >
        Test Title
      </h2>
    </a>
  </div>
</DocumentFragment>
`;
